### PR TITLE
Mail: added support for saving to eml and uploading eml

### PR DESCRIPTION
**Pull request by jkuchar**

### CODE REVIEW

File: MailLibrary/Drivers/IDriver.php [hunk 1]
```diff
@@ -134,10 +134,23 @@ function copyMail($mailId, $toMailbox);
...
 	function moveMail($mailId, $toMailbox);
 
+
+	/**
+	 * Save mail content as eml to given file
+	 * @param int $mailId
+	 * @param $file
+	 * @throws DriverException
+	 */
+	function saveMail($mailId, $file);
```
Comment (jkuchar) [inline MailLibrary/Drivers/IDriver.php]: path to file does not seem to me as good approach se we are assuming that e-mail is writable filesystem; I would use stream or just string with e-mail mime-message content.

File: MailLibrary/Drivers/IDriver.php [hunk 2]
```diff
@@ -134,10 +134,23 @@ function copyMail($mailId, $toMailbox);
...
+
+	function getHeaderInfo($mailId);
```
Comment (jkuchar) [inline MailLibrary/Drivers/IDriver.php]: What is this for? You can always use `$mail->getHeader(Mail::TO)` or `$mail[Mail::TO]` to get required headers. Is there anything I am missing?

File: MailLibrary/Drivers/ImapDriver.php [hunk 3]
```diff
@@ -487,4 +487,40 @@ protected function encodeMailboxName($name)
...
+		return $headers;
+	}
+
+	public function uploadMail($content)
+	{
+		if(!imap_append($this->resource, $this->server, $content)) {
+			throw new DriverException("Cannot upload mail: ".imap_last_error());
+		}
+
+	}
```
Comment (jkuchar) [inline MailLibrary/Drivers/ImapDriver.php]: I will rename 
- saveMail() -> retrieveRawMessage()
- uploadMail() -> uploadRawMessage()

File: MailLibrary/Mail.php [hunk 4]
```diff
@@ -248,6 +248,16 @@ public function delete()
...
 	}
 
+	public function saveToEml($file)
+	{
+		$this->connection->getDriver()->saveMail($this->id, $file);
+	}
+
+	public function getHeaderInfo()
+	{
+		return $this->connection->getDriver()->getHeaderInfo($this->id);
```
Comment (jkuchar) [inline MailLibrary/Mail.php]: You need to switch mailbox before each driver call as IMAP protocol works as access one mailbox per connection at a time.

**Comment by jkuchar** [comment]
pushed WIP version plese review

**Comment by jkuchar** [comment]
Idea, if we would pass Mailbox objects into $mail->move() and $mail->copy() we can simple move messages also across servers. For now as it is big change in API I would implement something like `MultiConnectionOperations::copy($message, $targetMailbox);` Do you agree?

**Comment by jkuchar** [comment]
added CrossConnectionOperations, the name need to be changes but I have no better idea for now. I would keep it WIP as implementation is very simple with notice (as is in header of file) that this functionality will be moved into Mail::move() and Mail::copy() later.

File: MailLibrary/Mail.php [hunk 5]
```diff
@@ -248,6 +248,7 @@ public function delete()
 		$this->connection->getDriver()->deleteMail($this->id);
 	}
 
+	/** @deprecated */
```
Comment (kerleba) [inline MailLibrary/Mail.php]: I think you can delete this. It is used only in my code and I will change it soon.

File: MailLibrary/Drivers/ImapDriver.php [hunk 6]
```diff
@@ -487,4 +487,61 @@ protected function encodeMailboxName($name)
 		return mb_convert_encoding($name, 'UTF7-IMAP', 'UTF-8');
 	}
 
+	/**
+	 * Save mail content as eml to given file
+	 * @param int $messageUID
+	 * @throws DriverException
+	 * @return string the mail content
+	 */
+	public function retrieveRawMessage($messageUID)
```
Comment (kerleba) [inline MailLibrary/Drivers/ImapDriver.php:496]: This is much better. I didn't realized that I can use `fopen('php://temp', 'w+b');`. Thank you.

File: MailLibrary/Drivers/IDriver.php (hunk 2 above)
Comment (kerleba) [reply MailLibrary/Drivers/IDriver.php]: I was not able to get date of delivery from this. So I used:
```
$headers = $mail->getHeaderInfo();
$headers->MailDate
```

File: MailLibrary/Mail.php (hunk 5 above)
Comment (jkuchar) [reply MailLibrary/Mail.php]: 👍

**Comment by jkuchar** [comment]
Getting delivery date

Problem: Delivery date is parameter of SMTP protocol and is known by target mail server. Original message does not contain such information. IMAP headers as they are more mail-server related, they contain such information.

IMAP headers are great place for checking that. However there is no way how to receive these information in maillib. I will create wrapping object for this, as it can be useful. Most clients first fetch header and then if they need more information they actualy starts downloading individual mime-parts...

Alternatively (the hard way) it is possible to extract delivery time from `Recieved:` headers, see following:

```
Return-Path: <bounce@detepe.sk>
X-Envelope-To: honza@kucharovi.cz
X-Spam-Level: 
Received: from mxload.webglobe.sk ([212.57.32.39])
	by grifart.cz (Kerio Connect 8.5.3)
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256 bits))
	for honza@kucharovi.cz;
	Fri, 6 Oct 2017 08:06:27 +0200
Received: from localhost (localhost [127.0.0.1])
	by mxload.webglobe.sk (Postfix) with ESMTP id C96D2195651A
	for <honza@kucharovi.cz>; Fri,  6 Oct 2017 08:06:24 +0200 (CEST)
X-Virus-Scanned: AVscan on mxload.webglobe.sk
Received: from mxload.webglobe.sk ([127.0.0.1])
	by localhost (inmail-06.webglobe.sk [127.0.0.1]) (amavisd-new, port 10024)
	with ESMTP id ZJEb2szZYObe for <honza@kucharovi.cz>;
	Fri,  6 Oct 2017 08:06:14 +0200 (CEST)
Received: from detepe.sk (wp-01.webglobe.sk [212.57.32.24])
	(Authenticated sender: newsletter@detepe.sk)
	by mxload.webglobe.sk (Postfix) with ESMTPSA id 40F3F195653B
	for <honza@kucharovi.cz>; Fri,  6 Oct 2017 08:06:07 +0200 (CEST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=detepe.sk; s=mail;
	t=1507269967; bh=ePTr66zXwBCxvmyySMFbyzCShZZuonFHcKXOj1xDK7I=;
	h=Date:To:From:Reply-To:Subject:List-Unsubscribe:From;
	b=RUOhNtT/WK3PuYB8qTPCfzFd4Bd5A4NrgPaXYiChU/T494wNId2c2u8SAdjRCMypk
	 258vKhD7VzcpdLhTrJkksqVQq6FEWjkzR5u09TnbEE8xsVtqd7Os+Ez85e2wR6ZlUY
	 79uUXo7k3i8ignTFbMJ33hmg7petw9sEaGhfUt6c=
Date: Fri, 6 Oct 2017 06:06:05 +0000
To: honza@kucharovi.cz
From: "DeTePe [dtp]" <newsletter@detepe.sk>
Reply-To: redakcia@detepe.sk
Subject: =?UTF-8?Q?nov=C3=BD_=C4=8Dl=C3=A1nok_na_DeTePe.sk_=E2=80=93_Font_d=C5=88a?=
 =?UTF-8?Q?_=E2=80=93_Fresh_Press?=
List-Unsubscribe: <https://detepe.sk/newsletter/unsubscribe>
MIME-Version: 1.0
Content-Type: multipart/alternative;
	boundary="b1_1d0887ae79d4130c8b79eaedb7550482"
Content-Transfer-Encoding: 8bit
```

However I will add IMAP headers object, as it is the simple way how to get to this information. These objects can be later used for lazy loading of messages, as some information is available just from them. 👍

**Comment by jkuchar** [comment]
So this looks like ready to merge... cfe754d

**Pull request merged**